src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -551,9 +551,23 @@ public function load(array $configs, ContainerBuilder $container)
         $container->registerForAutoconfiguration(LoggerAwareInterface::class)
             ->addMethodCall('setLogger', [new Reference('logger')]);
 
-        $container->registerAttributeForAutoconfiguration(AsEventListener::class, static function (ChildDefinition $definition, AsEventListener $attribute): void {
-            $definition->addTag('kernel.event_listener', get_object_vars($attribute));
-        });
+        if (\PHP_VERSION_ID >= 80000) {
+            $container->registerAttributeForAutoconfiguration(AsEventListener::class, eval(
+                <<<'PHP'
+                return static function (\Symfony\Component\DependencyInjection\ChildDefinition $definition, \Symfony\Component\EventDispatcher\Attribute\AsEventListener $attribute, \ReflectionClass|\ReflectionMethod $reflector): void {
```

Comment:
wrapping this in `eval` is pretty ugly. this is only due to the union type in `\ReflectionClass|\ReflectionMethod $reflector` right? why not just typhint `\Reflector $reflector` here? using a union type and sacrificing php 7 parsing for this does not seem worth it.

Reply:
The union type hint is done so that we can optimize. If we would add this on Reflector we'd have to loop through everything as we don't know what the attribute should work on.

Unless we use reflection on the attribute to find the target? And then use that information.

Reply:
Using the attribute target seems like to correct solution to me. Relying on the argument type is not reliable.

Reply:
Please explain why you think so? Because to me this is reliable.

Reply:
Because they would need to make sure the type is in sync with the attribute target themselves. If they use the wrong type, the configurator might just be ignored without even an error. Pretty bad DX to me.

Reply:
Same when using the attribute on eg private, protected methods when we read them only on public ones.
We cannot prevent every mistake, especially not when it'd mean doing extensive CPU processing.

Reply:
When would it be useful to have an attribute with class target and someone accidentlly typehints ReflectionMethod? There is no error or anything. Debugging hell.